Desktop #2383: Copy To Clipboard

Runo-saduwa opened the pull request:
### SCREENSHOT SHOWING THE BUTTON STYLE
![copy-paste-joplin](https://user-images.githubusercontent.com/35904727/74378445-442f7880-4de6-11ea-8a37-6ee9edb6ff22.PNG)

### TEST CASES
- Try with a code block - button appears
- Try with an inline code block - no button
- Try with a Fountain block - no button
- Try with a Katex block - no button
- Try with Katex inline code - no button
- Try with unformatted text (indented) that contains code - no button

laurent22 commented:
Thanks for the PR. Please install the precommit hook as described in BUILD.md as there are formatting issues.

I didn't have time to answer your questions in the original issue, but this feature would need to be developed as a markdown-it plugin. It allows the code to be more maintainable, by encapsulating the feature, and it means it's also less likely to interfere with other plugins.

It also mean that by developing it once it will work on both desktop on mobile.

So please could you repackage it as a plugin? You'll find some examples in the joplin-renderer directory, and you might also want to read the markdown-it doc to understand how the plugins work.

Runo-saduwa commented:
I will work on it again, thanks for the information!💪🏾

On Sat, Feb 1, 2020 at 3:25 PM Laurent Cozic <notifications@github.com>
wrote:

> Thanks for the PR. Please install the precommit hook as described in
> BUILD.md as there are formatting issues.
>
> I didn't have time to answer your questions in the original issue, but
> this feature would need to be developed as a markdown-it plugin. It allows
> the code to be more maintainable, by encapsulating the feature, and it
> means it's also less likely to interfere with other plugins.
>
> It also mean that by developing it once it will work on both desktop on
> mobile.
>
> So please could you repackage it as a plugin? You'll find some examples in
> the joplin-renderer directory, and you might also want to read the
> markdown-it doc to understand how the plugins work.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/laurent22/joplin/pull/2426?email_source=notifications&email_token=AIR5ZV6E65GCT6BPEICIMNLRAWA4JA5CNFSM4KOSOOU2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEKQ6IGQ#issuecomment-581035034>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIR5ZV7G3FIBP3BGOONH22TRAWA4JANCNFSM4KOSOOUQ>
> .
>

Runo-saduwa commented:
Ok here come again, how do I create a plugin?, I’ve learnt how to add a
plugin in to Markdown-it, but making plugin is the headache, I’ve looked
through the MdToHtml/rules directory which seems to be where these plugins
are created , but it’s not clear how I’ll build this feature

On Sat, Feb 1, 2020 at 3:31 PM Runo Saduwa <orunosaduwa@gmail.com> wrote:

> I will work on it again, thanks for the information!💪🏾
>
> On Sat, Feb 1, 2020 at 3:25 PM Laurent Cozic <notifications@github.com>
> wrote:
>
>> Thanks for the PR. Please install the precommit hook as described in
>> BUILD.md as there are formatting issues.
>>
>> I didn't have time to answer your questions in the original issue, but
>> this feature would need to be developed as a markdown-it plugin. It allows
>> the code to be more maintainable, by encapsulating the feature, and it
>> means it's also less likely to interfere with other plugins.
>>
>> It also mean that by developing it once it will work on both desktop on
>> mobile.
>>
>> So please could you repackage it as a plugin? You'll find some examples
>> in the joplin-renderer directory, and you might also want to read the
>> markdown-it doc to understand how the plugins work.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/laurent22/joplin/pull/2426?email_source=notifications&email_token=AIR5ZV6E65GCT6BPEICIMNLRAWA4JA5CNFSM4KOSOOU2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEKQ6IGQ#issuecomment-581035034>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AIR5ZV7G3FIBP3BGOONH22TRAWA4JANCNFSM4KOSOOUQ>
>> .
>>
>

laurent22 commented:
Have a look for example at the code_inline rule - it is used to add a class to a CODE element. You could duplicate this rule and instead process the code_block tokens. As a test, for example see if you can add a class to such a token.

Once you get this basic thing working, try to insert HTML elements within a code block (see the other rules to see how it's done). This is the key - once you managed to insert these extra elements, you can insert your button and position it the way you want with CSS.

See the "checkbox" rule to see how to add CSS from a rule.

But also check the markdown-it developer doc. It will tell you many details about how rules work: https://github.com/markdown-it/markdown-it/tree/master/docs

Runo-saduwa commented:
Thanks, I’ll do my best and get back to you as soon as possible

On Sat, Feb 1, 2020 at 7:12 PM Laurent Cozic <notifications@github.com>
wrote:

> Have a look for example at the code_inline rule - it is used to add a
> class to a CODE element. You could duplicate this rule and instead process
> the code_block tokens. As a test, for example see if you can add a class to
> such a token.
>
> Once you get this basic thing working, try to insert HTML elements within
> a code block (see the other rules to see how it's done). This is the key -
> once you managed to insert these extra elements, you can insert your button
> and position it the way you want with CSS.
>
> See the "checkbox" rule to see how to add CSS from a rule.
>
> But also check the markdown-it developer doc. It will tell you many
> details about how rules work:
> https://github.com/markdown-it/markdown-it/tree/master/docs
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/laurent22/joplin/pull/2426?email_source=notifications&email_token=AIR5ZV5TPPHHJUQQWEXQHB3RAW3RVA5CNFSM4KOSOOU2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEKRDFEA#issuecomment-581055120>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIR5ZV22VKGQ7GYM4C74AR3RAW3RVANCNFSM4KOSOOUQ>
> .
>

Runo-saduwa commented:
Hi @laurent22 , i am back again. i tried to follow your guide as much as possible, but some things still weren't quite clear, even the markdown-it docs is really unfriendly and i've done a lot of google but it seems like very few people have done stuff like this. but im really determined to solve this issue.

i have tried to implement the instructions you gave in the first paragraph, please kindly review and help me out with what needs to be changed and added next.

laurent22 commented:
That seems about right. You need to enable your rule in MdToHtml and then verify that it adds your class using the code inspector (in Help menu).

Runo-saduwa commented:
Alright, let me confirm that

Runo-saduwa commented:
@laurent22 i see it, Lol

i guess the next thing is to insert, elements inside the code block, can you please show me a really concrete and clear example to get this one done

laurent22 commented:
Best is to read through the markdown-it document again as this is where you'll find the most basic examples: https://github.com/markdown-it/markdown-it/blob/master/docs/architecture.md

Maybe also check our /rules directory, in particular the checkbox rule (but it's more complex).

Runo-saduwa commented:
work in progress, i will give updates on progress so far, its going fine. Thanks for the guidance

Runo-saduwa commented:
i have successfully developed the plugin.  I will still have to polish the styling but i will like the maintainers to test this branch before i finalize the styling. thank you

Runo-saduwa commented:
i have completed the styling, the PR is ready for review now

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml/copyFunction.js:
```diff
@@ -0,0 +1,35 @@
+(function(btn) {
+	const MDIC = {
```
Please use full words for function names and variables. `btn` = `button`. `MDIC` I don't know what it means.

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js:
```diff
@@ -0,0 +1,108 @@
+const fs = require('fs');
+const path = require('path');
+const filePath = path.join(__dirname, '../copyFunction.js');
+const btnCode = fs.readFileSync(filePath).toString();
```
That's an interesting approach to load the JS, but that won't work here because the "fs" package or reading file from the filesystem won't work on mobile (or browser, where this package can also be used). I think you can simply bring bring the copyFunction in this file directly.

Runo-saduwa replied:
done

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js:
```diff
@@ -0,0 +1,108 @@
+const copyCodeStyle = `
+.hljs {
+	position: relative;
+}
```
Why does the hljs class need to be changed? If it does need to be changed, it should in noteStyle.js directly.

Runo-saduwa replied:
i had to give it a ```position: relative```, to be able to to position the copy button with ```position:absolute```, but i'll take the style to noteStyle.js

Runo-saduwa replied:
change completed

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js:
```diff
@@ -0,0 +1,108 @@
+		successText = 'copied!', // copy-success text
+		successTextDelay = 2000, // successText show time [ms]
+	} = options;
+	const [tokens = {}, idx = 0] = args;
```
No default value please as the value should be there. If they aren't, you can throw an error (but probably not needed).

Runo-saduwa replied:
defaults removed

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js:
```diff
@@ -0,0 +1,108 @@
+	const copyTag = buttonBuilder.join('');
+
+	const originResult = render.apply(this, args);
+	const newResult = originResult.replace(LAST_TAG, `${copyTag}${LAST_TAG}`);
```
We shouldn't assume that the code block ends with a </PRE>. What if it's changed to a DIV later, or what if there are multiple </PRE> inside the block? (which I think can happen) Is there a more reliable way to insert this content without making assumption about how it ends?

Runo-saduwa replied:
i'll look into this

Runo-saduwa replied:
Good day @laurent22 , please can you suggest a way I could insert the
button into the <pre> element. This is my only challenge now

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js:
```diff
@@ -0,0 +1,108 @@
+	} = options;
+	const [tokens = {}, idx = 0] = args;
+	// get content
+	const cont = tokens[idx].content;
```
`cont` = `content`

Runo-saduwa replied:
done

laurent22 commented:
Thanks @Runo-saduwa, we're getting there and your approach seems correct overall. Did you try on the desktop app and does it work as expected?

I think we'll need to test the following cases on desktop:

- Try with a code block - button should appear
- Try with an inline code block - no button
- Try with a Fountain block - no button
- Try with a Katex block - no button
- Try with Katex inline code - no button
- Try with unformatted text (indented) that contains code - no button

See here for how to create these examples: https://joplinapp.org/markdown/

laurent22 commented:
Also please attach a screenshot to show how the button looks.

Runo-saduwa commented:
Good day @laurent22 , please can you suggest a way I could insert the
button into the <pre> element. This is my only challenge now

On Thu, Feb 6, 2020 at 11:11 AM Laurent Cozic <notifications@github.com>
wrote:

> *@laurent22* commented on this pull request.
> ------------------------------
>
> In
> ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js
> <https://github.com/laurent22/joplin/pull/2426#discussion_r375745551>:
>
> > +	const buttonBuilder = [
> +		'<div class="copy-wrapper">',
> +		`<div class="u-mdic-copy-notify" id="${uuid}">${successText}</div>`,
> +		'<button ',
> +		'class="u-mdic-copy-btn j-mdic-copy-btn" ',
> +		`data-mdic-content="${cont}" `,
> +		`data-mdic-notify-id="${uuid}" `,
> +		`data-mdic-notify-delay="${successTextDelay}" `,
> +		`data-mdic-copy-fail-text="${failText}" `,
> +		`onclick="${btnCode}"><i class="fa fa-copy"></i></button>`,
> +		'</div>',
> +	];
> +	const copyTag = buttonBuilder.join('');
> +
> +	const originResult = render.apply(this, args);
> +	const newResult = originResult.replace(LAST_TAG, `${copyTag}${LAST_TAG}`);
>
> We shouldn't assume that the code block ends with a . What if it's changed
> to a DIV later, or what if there are multiple inside the block? (which I
> think can happen) Is there a more reliable way to insert this content
> without making assumption about how it ends?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/laurent22/joplin/pull/2426?email_source=notifications&email_token=AIR5ZV57LEOGL6RIHMO2IADRBPO6PA5CNFSM4KOSOOU2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCUPLD5A#pullrequestreview-354333172>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIR5ZVZ5B6GUGMXR4DPHF4DRBPO6PANCNFSM4KOSOOUQ>
> .
>

laurent22 commented:
> Good day @laurent22 , please can you suggest a way I could insert the
> button into the <pre> element. This is my only challenge now

I don't know actually, as I haven't developed this type of plugin yet. You might need to experiment and try various things based on the documentation or existing examples.

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml.js:
```diff
@@ -11,12 +18,28 @@ const rules = {
+=======
+	copy: require('./MdToHtml/rules/copy_to_clipboard'),
+>>>>>>> finished up css
 };
```
That won't build.

laurent22 commented on ReactNativeClient/lib/joplin-renderer/assets/highlight.js/atom-one-light.css:
```diff
@@ -26,6 +26,7 @@ hue-6-2: #c18401
   background: #fafafa;
 }
 
+
```
Please remove this file from the commit.

laurent22 commented on package.json:
```diff
@@ -36,5 +36,8 @@
     "typescript": "^3.7.3"
+  },
+  "dependencies": {
+    "tesseract.js": "^2.0.2"
```
Unrelated to commit

laurent22 commented on package-lock.json:
```diff
@@ -364,6 +364,11 @@
       "integrity": "sha1-ibTRmasr7kneFk6gK4nORi1xt2c=",
```
Please remove this file from commit

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml.js:
```diff
@@ -2,6 +2,13 @@ const MarkdownIt = require('markdown-it');
+	failText: 'copy fail', // 'copy fail' | copy-fail text
+	successText: 'copied to clipboard!', // 'copy success' | copy-success text
+	successTextDelay: '2000', // 2000 | successText show time [ms]
+};
```
This structure appears in two locations, but you don't even need it. Please put the text and delay directly where it's needed.

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml.js:
```diff
@@ -161,7 +184,6 @@ class MdToHtml {
 					context.pluginAssets['highlight.js'] = [
 						{ name: options.codeTheme },
 					];
-
```
The commit would be much easier to read if it wasn't for all these unnecessary white space changes.

laurent22 commented:
Sorry too many problems with the commit at the moment, so I can't review. Please remove the unnecessary white space changes; remove the code unrelated to the PR, and please fix the multiple conflicts.

Runo-saduwa commented:
@laurent22 Thanks for the last review, i have made all the changes you asked. kindly review.

tessus commented:
The file `ReactNativeClient/lib/joplin-renderer/assets/highlight.js/atom-one-light.css` is still in your PR.

Runo-saduwa commented:
@tessus thanks for reminding me, i actually thought i got rid of that.

Runo-saduwa commented:
@laurent22 Good day, i’m waiting for feedback

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml.js:
```diff
@@ -11,6 +11,7 @@ const rules = {
 	highlight_keywords: require('./MdToHtml/rules/highlight_keywords'),
 	code_inline: require('./MdToHtml/rules/code_inline'),
 	fountain: require('./MdToHtml/rules/fountain'),
+	copy: require('./MdToHtml/rules/copy_to_clipboard'),
```
copy => copy_to_clipboard

Runo-saduwa replied:
Will do

laurent22 commented on ElectronClient/app/gui/note-viewer/index.html:
```diff
@@ -33,6 +34,25 @@
+
+#rendered-md pre:hover .copy{
+	display: block !important;
+} */
```
remove

laurent22 commented on ElectronClient/app/gui/note-viewer/index.html:
```diff
@@ -43,6 +63,28 @@
+	// })
+
+
+
```
remove

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js:
```diff
@@ -0,0 +1,137 @@
+}
+.copy-code-wrapper button i {
+font-size: 1rem;
+}
```
Please use tabs, not spaces for indentation.

Runo-saduwa replied:
Copy that

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js:
```diff
@@ -0,0 +1,137 @@
+		.replace(/>/g, '&gt;')
+		.replace(/'/g, '&apos;')
+		.replace(/"/g, '&quot;');
+}
```
Encoding/decoding doesn't work as expected. For example, I have this in my code block: `<b>M&amp;M's</b>` and press Copy, I want to get that exact text back, with no encoding/decoding. But with your current implementation I would get this back `<b>M&M's</b>`

Runo-saduwa replied:
I think I’ll have to remove this entire function because I found out markdown-it already handle encoding and decoding

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js:
```diff
@@ -0,0 +1,137 @@
+                    const copyButtonDataSet = _btn && _btn.dataset ? _btn.dataset : {};
+                    const notifyId = copyButtonDataSet.mdicNotifyId;
+                    const copyNotify = document.getElementById(notifyId);
+                    const notifyDelayTime = copyButtonDataSet.mdicNotifyDelay;
```
Please use tabs for indentation for all this.

Runo-saduwa replied:
Will do

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js:
```diff
@@ -0,0 +1,137 @@
+const copyStyle = `
+.code-n-btn-wrapper {
+  positon: relative;
```
"positon" is not a valid CSS property, and this is why your class "copy-code-wrapper" didn't work and you needed the hacky margins.

Runo-saduwa replied:
Thanks for showing me this!

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js:
```diff
@@ -0,0 +1,137 @@
+  position: absolute;
+  right: 0;
+  margin: 10px 15px 0 0;
+}
```
Once " .code-n-btn-wrapper" is fixed, please use top/right to position the button and remove the margin (which has a hidden dependency to the current theme).

Runo-saduwa replied:
Ok

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js:
```diff
@@ -0,0 +1,137 @@
+  font-size:0.9rem;
+  margin: 0 10px;
+  border-radius: 10px;
+}
```
Very light gray on white is very hard to read. Please use colour `#777` for instance.

Runo-saduwa replied:
Alright

laurent22 commented:
Still a bit more to do but we're nearly there. Also it doesn't work on mobile so please either make it work, or disable it in this context.

tessus commented:
I can help you fix this PR. But I need to know which files you actually modified.

If you post a list of the files you modified, I can have a look.

Runo-saduwa commented:
Thanks @tessus , Git stung me badly.

i Modified 
- `joplin\ReactNativeClient\lib\joplin-renderer\MdToHtml\rules\code_to_clipboard.js`
- `joplin\ReactNativeClient\lib\joplin-renderer\MdToHtml.js`

tessus commented:
ok, done.

Please learn how to use git. Check out topics like:

- rewrite git history
- deleting commits
- rebasing

Those are things that are often needed. The problem is, if you mess things up, you can lose your work, so you should really know how to use git.

Runo-saduwa commented:
i'm very grateful, i will work on my git skills

tessus commented:
You are welcome. Now you can address the issues Laurent mentioned.

tessus commented:
And do not do a git pull in your local branch !!!!!!!!!!!! Otherwise you mess it up again!

Runo-saduwa commented:
Ok. noted

tessus commented:
Delete your local branch and check it out from your remote.

Runo-saduwa commented:
Ok, will do

Runo-saduwa commented:
I have made the changes Laurent requested for which are:
- Changing some of the indentation to tabs
- changed copy to copy_to_clipboard in MdToHtml file
- The encoding/decoding works fine now. if you put anything like `<b>M&amp;M's</b>` in your block of code, you'll get the same thing back after copying
- corrected the typo `positon` to `position`
- removed the hacky margins and simply used `top` and 'right` for postioning
- changed grey color of notification to `#777`

Thats about it.

Runo-saduwa commented:
It seems like due to the fact that React Native has some  designated API for copy and pasting, i might not be able to use this present code to make copying to work on mobile, therefore we might just resolve to disabling the button in mobile but i dont really know how to go about it.

Please i'll love to hear your idea on how to go about this

tessus commented on ReactNativeClient/lib/joplin-renderer/MdToHtml.js:
```diff
@@ -200,15 +200,14 @@ class MdToHtml {
 		if (this.pluginEnabled('fountain')) markdownIt.use(rules.fountain(context, ruleOptions));
 		if (this.pluginEnabled('mermaid')) markdownIt.use(rules.mermaid(context, ruleOptions));
-		markdownIt.use(rules.sanitize_html(context, ruleOptions));
 		markdownIt.use(rules.highlight_keywords(context, ruleOptions));
```
Why is the `sanitize_html` deleted? Please add it back. That's probably the reason why the tests fail.

Runo-saduwa replied:
I don’t know how that happened but I’ll add it back anyway

tessus commented:
> It seems like due to the fact that React Native has some designated API for copy and pasting, i might not be able to use this present code to make copying to work on mobile, therefore we might just resolve to disabling the button in mobile but i dont really know how to go about it.

I don't underatand what you are saying. Why should this not work on mobile? Did you test it?

Runo-saduwa commented:
When Laurent reviewed this PR, he said it doesn’t work for mobile, then he said I should either make it work or disable it based on that context. 

I haven’t been able to build the mobile app on my machine because I don’t have Android studio, so I can’t test for mobile. 

So in conclusion, I’m trying to disable this plugin for mobile

Runo-saduwa commented:
Is there a way to build the mobile app using expo?, because that’s my last resort atm

tessus commented:
No idea, I just installed Android Studio and went from there.

You still haven't added thta one line. How long does it take to add one line?
We need to see, if this fixes the test error.

tessus commented:
We are reaching almost 80 comments. Not even a bug report had these many comments so far.
Maybe we should just forget about this and you open a new PR, when you are done with the rest. Don't get me wrong, but you should figure the last part out by yourself. Read some documentation, post in forums, or whatnot. If we tell you what to do all the time, we don't need the PR. How do you ever expect to work on a GSoC project that spans 2 months?

Runo-saduwa commented:
@tessus You’ve been of great help, I will take your advice and simply close this PR and come back stronger with a complete version. I will deliver
